Add formatted output generation on top of BufferWriter.

SolidWallOfCode opened the pull request.

Commit:
x-debug log-headers propagation and documentation

Now the log-headers request will propagate upstream even when logging is not enabled on the local box. This makes it easier to use on higher tier caches. 

Also added documentation.

SolidWallOfCode committed:
Split MemSpan from TextView, cleanup MemSpan API, add documentation.

SolidWallOfCode committed:
bwprint: Type safe printf like output to BufferWriter instances.

PSUdaemon commented on lib/ts/bwprint.cc:
```diff
@@ -0,0 +1,410 @@
+      prefix = "0B"_sv;
+    n        = detail::bwf_to_radix<2>(i, buff, sizeof(buff), detail::UPPER_DIGITS);
+    break;
+    break;
```
Extra break?

SolidWallOfCode replied:
Dang, I thought I'd fixed that.

PSUdaemon commented on lib/ts/bwprint.cc:
```diff
@@ -0,0 +1,410 @@
+  case 'x':
+    if (spec._radix_lead_p)
+      prefix = "0x"_sv;
+    n        = detail::bwf_to_radix<16>(i, buff, sizeof(buff), detail::LOWER_DIGITS);
```
Putting braces around the body of the `if` will fix this weird formatting and also match our style guide.

SolidWallOfCode replied:
Haha, yeah, that's some serious ugly there, isn't it?

ywkaras commented on lib/ts/bwprint.cc:
```diff
@@ -0,0 +1,410 @@
+          if (!sz)
+            return;
+        } else {
+          throw std::invalid_argument("Precision mark without precision");
```
Generic use of exceptions is questionable in a source base that has not be upgraded to exception safety.

ywkaras commented on lib/ts/bwprint.cc:
```diff
@@ -0,0 +1,410 @@
+
+ts::BWFormat::~BWFormat()
+{
+}
```
Why?  Just, why?

SolidWallOfCode replied:
Habit.

SolidWallOfCode commented:
I really don't like the duplication between `bwprint` and `BWFormat::BWFormat`. I need to figure out how to unify those two things.

SolidWallOfCode commented:
[approve ci]

Commit:
Framework for CPPAPI regression test plugin run by a gold test.

dolladollabill committed:
add custom --help and -h flag to traffic_ctl

Commit:
Remove inefficient use of std::string in C++ transformation plugins.

Change 'const std::string &' parameters to virtual member functions of the TranformationPlugin class to ts::string_view.  This
allows derived classes of TranformationPlugin to avoid using std::string in some cases, resulting in less copying and dynamic
memory allocation.

shukitchan committed:
Support TSRemapFromUrlGet/TSRemapToUrlGet in ts_lua plugin

SolidWallOfCode committed:
Doc: Update stripe assignment initialization documentation.

shinrich committed:
Stop connection reattempts on origin which sends EOS on no data.

SolidWallOfCode committed:
Errata: CacheTool updates.

shinrich committed:
Clear vio so lingering IO is not confused when error message is returned

Addresses issue #3239

shinrich committed:
Syntax check should ignore age

The same element is fetched in test one and 5.  On a loaded system it
could be in cache for a second and report age:1 instead of age:0

Commit:
Remove leap-into-the-abyss casting from TSTransformCreate().

Commit:
Fix clang format errors.

scw00 committed:
Correct shutdown flag in UnixNetVConnection

randall committed:
Removes the last remnants of tsqa tests

Commit:
corrects doc example for s3_auth map rule

Commit:
Fixing 404 on redirects

This is a bug introduced with PR #2335. The `if` is to avoid unnecessary
works when buffering the post body, but since redirects also need to
execute the logic flow and the request info may be changed during the
process of redirects. We need to extract the request info every time redirects
happen.

shukitchan committed:
remove unnecessary use of snprintf

SolidWallOfCode committed:
Scalar: Updates and convert tests to Catch.
